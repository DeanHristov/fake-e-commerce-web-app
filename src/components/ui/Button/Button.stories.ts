import type {Meta, StoryObj} from '@storybook/react';

import Button, {IButtonProps} from './Button'; //@see:  https://storybook.js.org/docs/react/writing-stories/introduction

//@see:  https://storybook.js.org/docs/react/writing-stories/introduction
const meta: Meta<IButtonProps> = {
  title: 'Component/Button',
  component: Button,
  parameters: {
    // Optional parameter to center the component in the Canvas. More info: https://storybook.js.org/docs/react/configure/story-layout
    layout: 'centered',
  },
  argTypes: {},
};

type Story = StoryObj<IButtonProps>;

// More on writing stories with args: https://storybook.js.org/docs/react/writing-stories/args
export const Primary: Story = {
  args: {
    title: 'Click me',
  },
};

export default meta;
