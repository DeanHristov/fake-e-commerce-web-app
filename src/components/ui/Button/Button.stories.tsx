import type {Meta, StoryObj} from '@storybook/react';
import {ShoppingCartIcon} from '@heroicons/react/24/solid';

import Button from './Button'; //@see:  https://storybook.js.org/docs/react/writing-stories/introduction

//@see:  https://storybook.js.org/docs/react/writing-stories/introduction
const meta: Meta<typeof Button> = {
  title: 'Component/Button',
  component: Button,
  parameters: {
    // Optional parameter to center the component in the Canvas. More info: https://storybook.js.org/docs/react/configure/story-layout
    layout: 'centered',
  },

  argTypes: {
    onClick: { action: 'You clicked me!' },
    variant: {
      options: ['primary', 'secondary'],
      control: { type: 'select' },
    },
  },
};

type Story = StoryObj<typeof Button>;

export const Primary: Story = {
  args: {
    // variant: 'primary',
    title: 'Click me',
    spinning: false,
  },
};

export const WithLeftIcon: Story = {
  args: {
    title: 'Click me',
    spinning: false,
    leftIcon: <ShoppingCartIcon className="h-6 w-6 text-white" />,
  },
};

export const WithRightIcon: Story = {
  args: {
    title: 'Click me',
    spinning: false,
    rightIcon: <ShoppingCartIcon className="h-6 w-6 text-white" />,
  },
};

export default meta;
