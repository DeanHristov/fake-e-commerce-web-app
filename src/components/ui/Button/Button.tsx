'use client';

import { FC, ReactNode } from 'react';
import Spinner from '@/components/ui/Spinner';

export type TButtonPresentations = 'primary' | 'secondary';
export type TButtonType = 'submit' | 'reset' | 'button';

export interface IButtonProps {
  title?: string;
  type?: TButtonType;
  className?: string;
  disabled?: boolean;
  spinning?: boolean;
  variant?: TButtonPresentations;
  onClick?: () => void;
  leftIcon?: ReactNode;
  rightIcon?: ReactNode;
}

const Button: FC<IButtonProps> = ({
  variant = 'primary',
  title,
  leftIcon,
  rightIcon,
  spinning,
  className,
  disabled = false,
  onClick = () => {},
  ...rest
}) => {
  const buttonStyles: Record<string, string> = {
    primary:
      'bg-green-600 text-white font-bold uppercase bg-primary hover:bg-hover',
    secondary: 'bg-white border border-green-600 text-dark font-normal',
  };

  return (
    <button
      disabled={disabled}
      className={`${
        disabled ? 'cursor-not-allowed opacity-50' : 'cursor-pointer'
      } ${
        buttonStyles[variant]
      } hover:shadow-lg transition transform active:scale-90 duration-200 ease-in w-full px-4 py-1.5 flex justify-center items-center gap-2 ${className}`}
      onClick={onClick}
      {...rest}
    >
      {leftIcon}
      {title}
      {rightIcon}
      {spinning && <Spinner data-testid="spinner" width={24} color="#155e75" />}
    </button>
  );
};

export default Button;
