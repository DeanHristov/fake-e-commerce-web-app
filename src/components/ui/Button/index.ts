import Button, { IButtonProps } from './Button';

export type { IButtonProps };

export default Button;
