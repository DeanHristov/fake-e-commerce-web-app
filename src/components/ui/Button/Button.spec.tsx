import { render, waitFor } from '@testing-library/react';
import Button from './Button';
import userEvent from '@testing-library/user-event';

describe('UI/Component <Button {...} />', () => {
  it('Should render default state', () => {
    const { container } = render(<Button title="Click me" />);

    expect(container).toMatchSnapshot();
  });

  it('Should be able to trigger click event', async () => {
    const onChangeMock = jest.fn();
    const { getByRole } = render(
      <Button title="Click me" onClick={onChangeMock} />,
    );

    const button = getByRole('button', { name: /Click me/i });
    await waitFor(() => userEvent.click(button));

    expect(onChangeMock).toBeCalled();
  });
});
