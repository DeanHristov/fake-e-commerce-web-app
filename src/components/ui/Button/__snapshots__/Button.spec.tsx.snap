// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`UI/Component <Button {...} /> Should be able to render it with a  left icon 1`] = `
<div>
  <button
    class="cursor-pointer bg-green-600 text-white font-bold uppercase bg-primary hover:bg-hover hover:shadow-lg duration-300 ease-in w-full px-4 py-1.5 flex justify-center items-center gap-2 rounded-md"
  >
    <svg
      aria-hidden="true"
      class="h-6 w-6 text-white"
      fill="currentColor"
      viewBox="0 0 24 24"
      xmlns="http://www.w3.org/2000/svg"
    >
      <path
        d="M2.25 2.25a.75.75 0 000 1.5h1.386c.17 0 .318.114.362.278l2.558 9.592a3.752 3.752 0 00-2.806 3.63c0 .414.336.75.75.75h15.75a.75.75 0 000-1.5H5.378A2.25 2.25 0 017.5 15h11.218a.75.75 0 00.674-.421 60.358 60.358 0 002.96-7.228.75.75 0 00-.525-.965A60.864 60.864 0 005.68 4.509l-.232-.867A1.875 1.875 0 003.636 2.25H2.25zM3.75 20.25a1.5 1.5 0 113 0 1.5 1.5 0 01-3 0zM16.5 20.25a1.5 1.5 0 113 0 1.5 1.5 0 01-3 0z"
      />
    </svg>
    Click me
  </button>
</div>
`;

exports[`UI/Component <Button {...} /> Should be able to render it with a right icon 1`] = `
<div>
  <button
    class="cursor-pointer bg-green-600 text-white font-bold uppercase bg-primary hover:bg-hover hover:shadow-lg duration-300 ease-in w-full px-4 py-1.5 flex justify-center items-center gap-2 rounded-md"
  >
    Click me
    <svg
      aria-hidden="true"
      class="h-6 w-6 text-white"
      fill="currentColor"
      viewBox="0 0 24 24"
      xmlns="http://www.w3.org/2000/svg"
    >
      <path
        d="M2.25 2.25a.75.75 0 000 1.5h1.386c.17 0 .318.114.362.278l2.558 9.592a3.752 3.752 0 00-2.806 3.63c0 .414.336.75.75.75h15.75a.75.75 0 000-1.5H5.378A2.25 2.25 0 017.5 15h11.218a.75.75 0 00.674-.421 60.358 60.358 0 002.96-7.228.75.75 0 00-.525-.965A60.864 60.864 0 005.68 4.509l-.232-.867A1.875 1.875 0 003.636 2.25H2.25zM3.75 20.25a1.5 1.5 0 113 0 1.5 1.5 0 01-3 0zM16.5 20.25a1.5 1.5 0 113 0 1.5 1.5 0 01-3 0z"
      />
    </svg>
  </button>
</div>
`;

exports[`UI/Component <Button {...} /> Should be able to render the primary state 1`] = `
<div>
  <button
    class="cursor-pointer bg-green-600 text-white font-bold uppercase bg-primary hover:bg-hover hover:shadow-lg duration-300 ease-in w-full px-4 py-1.5 flex justify-center items-center gap-2 rounded-md"
  >
    Click me
  </button>
</div>
`;

exports[`UI/Component <Button {...} /> Should be able to render the secondary state 1`] = `
<div>
  <button
    class="cursor-pointer bg-white border border-green-600 text-dark font-normal hover:shadow-lg duration-300 ease-in w-full px-4 py-1.5 flex justify-center items-center gap-2 rounded-md"
  >
    Click me
  </button>
</div>
`;
